### ISSUE TITLE
tqdm progress bars break when used inside objectives

### ISSUE THREAD

**Issue by nalourie-ai2** [hyperopt/hyperopt#476]
When an objective function uses a tqdm progress bar, the output is not properly removed from the screen. The bug occurs both when `fmin` is used with `show_progressbar=False` as well as when it's used with `show_progressbar=True`. 

I obtained this issue using hyperopt version `0.1.2`. The issue doesn't occur in version `0.1.1`.

Here is my python environment:

```
decorator==4.3.2
future==0.17.1
hyperopt==0.1.2
networkx==2.2
numpy==1.16.2
pymongo==3.7.2
scipy==1.2.1
six==1.12.0
tqdm==4.31.1
```

And here is a minimal example to reproduce the issue:

```python
import time

from hyperopt import fmin, hp, tpe
import tqdm

def objective(x):
    for _ in tqdm.tqdm(range(2)):
        time.sleep(0.5)
    return x


print('\n\n#### Bug with ``show_progressbar=True`` ####\n\n')

fmin(
    fn=objective,
    space=hp.uniform('x', 0, 1),
    algo=tpe.suggest,
    max_evals=3,
    show_progressbar=True)


print('\n\n#### Bug with ``show_progressbar=False`` ####\n\n')

fmin(
    fn=objective,
    space=hp.uniform('x', 0, 1),
    algo=tpe.suggest,
    max_evals=3,
    show_progressbar=False)
```

I believe this issue is caused by `std_out_err_redirect_tqdm` context manager, though I have not investigated enough to verify whether or not that's true.

**Comment by PhilipMay** [comment]
Ohh that's bad. Sorry. I will investigate this. I hope you found a workaround.

**Comment by maxpumperla** [comment]
@PhilipMay that was precisely the question I had for you before merging that PR. It's a little unfortunate, but not that big of a deal. If you can please provide a fix relatively soon, that'd be great. Otherwise I'll have to revert the PR and cut a new minor release.

**Comment by PhilipMay** [comment]
Well. Redirect writing (https://github.com/tqdm/tqdm#redirecting-writing) combined with nested progress bars (https://github.com/tqdm/tqdm#nested-progress-bars) does not seem to work. Either you do not redirect and break the progressbar when you do any `print()` output or you redirect and that breaks the nesting. I can not provide a fix or improvement for this. I am sorry. :-(

**Comment by david-waterworth** [comment]
I also had issues with this because my objective calls a docker sdk container and writes the docker log back to stdout. `std_out_err_redirect_tqdm` causes this to fail (I replaced `sys.stdout.buffer.write(line)` with `print(line.decode("utf-8").rstrip())`

            cmd = f'train --model "{model}" --site {site_id} --streams ' + ' '.join(f'"{s}"' for s in streams)
            container = client.containers.run(image, command=cmd, runtime='nvidia', volumes = {mount:{'bind': '/opt/ml', 'mode': 'rw'}}, detach=True)
            for line in container.logs(stream=True):
                sys.stdout.buffer.write(line)

**Comment by maxpumperla** [comment]
@PhilipMay there's no way to silence the progress bar by default? I don't have time to look into this right now, but I'd rather fix this feature than revert it. The `show_progressbar=False` case should behave as in previous hyperopt versions. Why is that not possible?

**Comment by david-waterworth** [comment]
Note, I fixed my issue after reading https://github.com/tqdm/tqdm#writing-messages

Now I can output the docker log to the console inside my fmin objective.

**Comment by artyompal** [comment]
What is the reason for `std_out_err_redirect_tqdm()` call to exist? It breaks `tqdm()` invocations in inner calls. It turns 1 line of console output into a thousand, making output a mess. Imho, it shouldn't be there at all, or it should be configurable.

In general, I think that the context manager is an antipattern here. Libraries should not tweak application code.

**Comment by artyompal** [comment]
Please, can we have an option to disable std_out_err_redirect_tqdm()?  It's really annoying since it breaks all tqdm calls in the callback. 

Logically, it shouldn't be there if show_progressbar=False.

**Comment by easyas314159** [comment]
This starts to become a real issue when using hyperopt in combination with Keras for neural networks with a large number of training batches as it makes the output effectively unreadable.

There are a couple of solutions to this however I was looking for some feedback before putting together a PR with a fix.

Option 1: Remove `std_out_err_redirect_tqdm` and stop overriding the `file` argument to `tqdm`. This would resolve the issue however it would cause `print` output within the function being evaluated to potentially cause weird display behavior (See: https://github.com/hyperopt/hyperopt/pull/455#issuecomment-454567396)

Option 2: Wrap the `std_out_err_redirect_tqdm` and `tqdm` call with-in `contextlib.ExitStack` so that if `show_progress` is set to `False` then the progress bar is never created and will not interfere with child progress bars. Unfortunately using `contextlib.ExitStack` breaks python 2.7 support, however, with support for 2.7 ending in a month is this even an issue? In addition, this removes the overall progress bar which I would still like to preserve.

Option 3: Remove the progress bar from PR #455 and replace it with an optional progress callback. This provides slightly more fine-grained control to the user and hyperopt could include a default progress callback that would preserve the existing functionality. I am not quite sure what the architecture for this would look like yet but a little more thought should solve that problem.

Does anyone have any additional thoughts on other possible solutions, that preserve the existing good behaviour with `print` and also behaves well with nested `tqdm` objects?

**Comment by PhilipMay** [comment]
@easyas314159 thanks - maybe you could review this PR: https://github.com/hyperopt/hyperopt/pull/583 

Does it fix the problem?

**Comment by easyas314159** [comment]
@PhilipMay It does solve the underlying problem with `std_out_err_redirect_tqdm` and fixes the issue as it presents in a terminal window.

I suppose this potentially leads into a broader discussion about progress monitoring of the optimization process. For example, there is currently no easy way to cancel the optimization process without access to the `Trials` object, or a mechanism for monitoring no improvement in a while, or even monitoring losses outside the main objective. A more robust callback system in general with hooks into various steps of the optimization process would be extremely useful.

Something kind of like this comes to mind:
```python
class Callback:
	def on_optimization_start(self, trials):
		...

	def on_optimization_end(self, trials):
		...

	def on_trial_start(self, trials, trial):
		...

	def on_trial_end(self, trials, trial):
		...
```

I can think of a handful of things this could be used for:
- Early stopping when there is insufficient improvement
- Early stopping when a threshold is passed
- Monitoring other properties of the trial object that are not the loss
- Custom progress monitoring that sends status information via HTTP to another service
- Custom progress bars with options for more than just a `best loss: %f` postfix

I can technically do all of that myself but it requires calling `fmin` for a single iteration and handling all externally which is not a great option in a distributed context.

This should probably be opened as a separate issue/enhancement? I also don't mind taking the development of this on but it will have to wait until later this month,

**Comment by PhilipMay** [comment]
@easyas314159 Thanks for your feedback. IMO your ideas are great. I would suggest to start a new issue to discuss them.

**Comment by maxpumperla** [comment]
@easyas314159 @PhilipMay I'd love to get a discussion started around this and plan/design this with you. Callbacks and early stopping have been on my bucket list for ages, your help is very much appreciated.

**Comment by PhilipMay** [comment]
@maxpumperla @easyas314159 started a dedicated issue here: #585

**Comment by PhilipMay** [comment]
#583 is merged, new issue for callbacks is started and this should be fixed. 
Can it be closed?
